### PR TITLE
[coro][pgo] Don't promote pgo counters in the suspend basic block

### PR DESCRIPTION
**Pull request by mtrofin**
If a suspend happens in the resume part (this can happen in the case of
chained coroutines), and that's part of a loop, the pre-split CFG has
the suspend block as an exit of that loop. PGO Counter Promotion will
then try to commit the temporary counter to the global in that "exit"
block (it also does that in the other loop exit BBs, which also includes
the "destroy" case).

We don't need to commit the counter in the suspend case - it's not
a loop exit from the perspective of the behavior of the program. The
regular loop exit, together with the "destroy" case, completely cover
any updates that may need to happen to the global counter.

### CODE REVIEW

File: llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp [hunk 1]
```diff
@@ -242,8 +242,16 @@ class PGOCounterPromoter {
     if (!isPromotionPossible(&L, LoopExitBlocks))
       return;
 
+    auto IsSuspendBB = [&](BasicBlock *BB) {
```
Comment (david-xl) [inline llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:245]: can this be a common utility function somewhere such as BasicBlockUtils.h?

File: llvm/test/Transforms/Coroutines/coro-split-musttail-chain-pgo-counter-promo.ll [hunk 2]
```diff
@@ -0,0 +1,175 @@
+; REQUIRES: x86-registered-target
+; RUN: opt -passes='pgo-instr-gen,instrprof,coro-split' -do-counter-promotion=true -S < %s | FileCheck %s
```
Comment (david-xl) [inline llvm/test/Transforms/Coroutines/coro-split-musttail-chain-pgo-counter-promo.ll:2]: the test case is a little large. Possible to reduce it ?

File: llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp (hunk 1 above)
Comment (mtrofin) [reply llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:245]: This is a perfect segway to whether we should handle the suspend/switch pattern somewhere more centrally - I wouldn't be surprised other analyses would make incorrect inferences about the 'suspend (i.e. ret)' case.

Comment (ChuanqiXu9) [reply llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:245]: In fact, there are other passes failed to handle the suspend intrinsic. Originally we were wondering if we can fix the issues by providing a new analysis pass (e.g., CoroInfoPass). But we failed to go on (due to limited development resources). So it sounds a good idea to extract the common patterns to common utilities. Although I don't think this should be a requirement for this specific PR.

**Comment by mtrofin** [comment]
> In fact, there are other passes failed to handle the suspend intrinsic. Originally we were wondering if we can fix the issues by providing a new analysis pass (e.g., CoroInfoPass). But we failed to go on (due to limited development resources). So it sounds a good idea to extract the common patterns to common utilities. Although I don't think this should be a requirement for this specific PR.  

Ack. Is there a Discourse topic on this, otherwise I can start one. We could, for example, add a IR validation clause, too.

File: llvm/include/llvm/Transforms/Utils/BasicBlockUtils.h [hunk 3]
```diff
@@ -705,6 +705,8 @@ void InvertBranch(BranchInst *PBI, IRBuilderBase &Builder);
 // Check whether the function only has simple terminator:
 // br/brcond/unreachable/ret
 bool hasOnlySimpleTerminator(const Function &F);
+
+bool isPresplitCoroSuspendExit(const BasicBlock &BB);
```
Comment (david-xl) [inline llvm/include/llvm/Transforms/Utils/BasicBlockUtils.h:709]: should the minimal spanning tree code use this interface too?

**Comment by ChuanqiXu9** [comment]
> > In fact, there are other passes failed to handle the suspend intrinsic. Originally we were wondering if we can fix the issues by providing a new analysis pass (e.g., CoroInfoPass). But we failed to go on (due to limited development resources). So it sounds a good idea to extract the common patterns to common utilities. Although I don't think this should be a requirement for this specific PR.
> 
> Ack. Is there a Discourse topic on this, otherwise I can start one. We could, for example, add a IR validation clause, too.

As far as I know, we don't have such Discourse topic. It is seperated in previous commit review process.

**Comment by mtrofin** [comment]
> should the minimal spanning tree code use this interface too? 

Good idea - done. It needed a little rework - in the MST case we want to work with edges and block splitting the suspend faux edge. In the coro promotion case, we care about the basic block, and the property that none of the incoming edges is a faux suspend edge.

**Comment by mtrofin** [comment]
> > > In fact, there are other passes failed to handle the suspend intrinsic. Originally we were wondering if we can fix the issues by providing a new analysis pass (e.g., CoroInfoPass). But we failed to go on (due to limited development resources). So it sounds a good idea to extract the common patterns to common utilities. Although I don't think this should be a requirement for this specific PR.
> > 
> > 
> > Ack. Is there a Discourse topic on this, otherwise I can start one. We could, for example, add a IR validation clause, too.
> 
> As far as I know, we don't have such Discourse topic. It is seperated in previous commit review process.

Started https://discourse.llvm.org/t/coro-pre-split-handling-of-the-suspend-edge/75043

File: llvm/include/llvm/Transforms/Utils/BasicBlockUtils.h (hunk 3 above)
Comment (mtrofin) [reply llvm/include/llvm/Transforms/Utils/BasicBlockUtils.h]: Good idea - done. It needed a little rework - in the MST case we want to work with edges and block splitting the suspend faux edge. In the coro promotion case, we care about the basic block, and the property that none of the incoming edges is a faux suspend edge.

**Review by david-xl — APPROVED** [review]

File: llvm/include/llvm/Transforms/Utils/BasicBlockUtils.h [hunk 4]
```diff
@@ -705,6 +705,9 @@ void InvertBranch(BranchInst *PBI, IRBuilderBase &Builder);
 // Check whether the function only has simple terminator:
 // br/brcond/unreachable/ret
 bool hasOnlySimpleTerminator(const Function &F);
+
+bool isPresplitCoroSuspendExitEdge(const BasicBlock &Src,
```
Comment (ChuanqiXu9) [inline llvm/include/llvm/Transforms/Utils/BasicBlockUtils.h:709]: nit: Let's add some comments for this. Especially all other function in Utils.h have a lot of comments.

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM with a nit.

**Pull request merged**